src/monitor/group_state_machine.c

```diff
@@ -82,27 +87,42 @@ ProceedGroupState(AutoFailoverNode *activeNode)
 		return true;
 	}
 
-	/* single -> wait_primary when another node wants to become standby */
-	if (IsCurrentState(activeNode, REPLICATION_STATE_SINGLE) &&
-		IsCurrentState(otherNode, REPLICATION_STATE_WAIT_STANDBY))
+	/*
+	 * We separate out the FSM for the primary server, because that one needs
+	 * to loop over every other node to take decisions. That induces some
+	 * complexity that is best managed in a specialized function.
+	 */
+	if (IsInPrimaryState(activeNode))
 	{
-		char message[BUFSIZE];
+		return ProceedGroupStateForPrimaryNode(activeNode);
+	}
```

Comment:
Would it make sense to put the remainder of the function in a ProceedGroupStateForSecondaryNode?

Reply:
I though about it, and decided to minimise the changes as much as I could, so that we can see the same code as before in the same function, but with more cases added to it.

In the final PR we have the following code for multiple standby failover orchestration for instance:

``` c
	/* Multiple Standby failover is handled in its own function */
	if (IsUnhealthy(primaryNode) && nodesCount > 2)
	{
		/*
		 * ProceedGroupStateForMSFailover chooses the failover candidate when
		 * there's more than one standby node around, by applying the
		 * candidatePriority and comparing the reportedLSN. The function also
		 * orchestrate fetching the missing WAL from the failover candidate if
		 * that's needed.
		 *
		 * When ProceedGroupStateForMSFailover returns true, it means it was
		 * successfull in driving the failover to the next step, and we should
		 * stop here. When it return false, it did nothing, and so we want to
		 * apply the common orchestration code for a failover.
		 */
		if (ProceedGroupStateForMSFailover(activeNode, primaryNode))
		{
			return true;
		}
	}
```

So it could make sense to have an even more explicit split, maybe, yes.